src/engine/Engine.cpp

```diff
@@ -3106,4 +3202,391 @@ InputQuery Engine::buildQueryFromCurrentState() const {
+
+    unsigned infeasibleVar = _boundManager.getInconsistentVariable();
+
+    if ( infeasibleVar == IBoundManager::NO_VARIABLE_FOUND || !certifyInfeasibility( infeasibleVar ) )
+        infeasibleVar = explainFailureWithTableau();
+
+    if ( infeasibleVar == IBoundManager::NO_VARIABLE_FOUND )
+        infeasibleVar = explainFailureWithCostFunction();
+
+    if ( infeasibleVar == IBoundManager::NO_VARIABLE_FOUND )
+    {
+        _costFunctionManager->computeCoreCostFunction();
+        infeasibleVar = explainFailureWithCostFunction();
+    }
+
+    if ( infeasibleVar == IBoundManager::NO_VARIABLE_FOUND )
+    {
+        markLeafToDelegate();
+        return;
+    }
+
+    ASSERT( infeasibleVar < _tableau->getN() );
+    ASSERT( _UNSATCertificateCurrentPointer && !( **_UNSATCertificateCurrentPointer ).getContradiction() );
+    _statistics.incUnsignedAttribute( Statistics::NUM_CERTIFIED_LEAVES );
+
+    writeContradictionToCertificate( infeasibleVar );
+
+    ( **_UNSATCertificateCurrentPointer ).makeLeaf();
+}
+
+bool Engine::certifyInfeasibility( unsigned var ) const
+{
+   ASSERT( _produceUNSATProofs );
+
+    Vector<double> contradiction = computeContradiction( var );
+
+    if ( contradiction.empty() )
+        return FloatUtils::isNegative( _groundBoundManager.getUpperBound( var ) - _groundBoundManager.getLowerBound( var ) );
+
+    double derivedBound = UNSATCertificateUtils::computeCombinationUpperBound( contradiction.data(), _tableau->getSparseA(),
+                                                                               _groundBoundManager.getUpperBounds(), _groundBoundManager.getLowerBounds(),
+                                                                               _tableau->getM(), _tableau->getN() );
+    return FloatUtils::isNegative( derivedBound );
+}
+
+double Engine::explainBound( unsigned var, bool isUpper ) const
+{
+    ASSERT( _produceUNSATProofs );
+
+    Vector<double> explanationVec( 0 );
+
+    if  ( !_boundManager.isExplanationTrivial( var, isUpper ) )
+        _boundManager.getExplanation( var, isUpper, explanationVec );
+
+    const double *explanation = explanationVec.empty() ? NULL : explanationVec.data();
+    return UNSATCertificateUtils::computeBound( var, isUpper, explanation, _tableau->getSparseA(),
+                                                _groundBoundManager.getUpperBounds(), _groundBoundManager.getLowerBounds(),
+                                                _tableau->getM(), _tableau->getN() );
+}
+
+bool Engine::validateBounds( unsigned var, double epsilon, bool isUpper ) const
+{
+    ASSERT( _produceUNSATProofs );
+
+    double explained, real;
+    explained = explainBound( var, isUpper );
+    if ( isUpper )
+    {
+        real = _boundManager.getUpperBound( var );
+        if ( explained - real > epsilon )
+        {
+            printf( "Var %d. Computed Upper %.5lf, real %.5lf. Difference is %.10lf\n", var, explained, real, abs( explained - real ) );
```

Comment:
It is a debugging function which is very useful for checking bound explanations are correct. It is not currently used, not even in debug mode, since it significantly reduces the speed.